### PR TITLE
Fix dynamic zone id conflicts and editor re-rendering

### LINKED ISSUE #17858: Cannot update a richtext field after sibling deletion within dynamic components
### Required System information

- Node.js version: v18.17.1
- NPM version: 8.19.3
- Strapi version: v4.12.7
- Is your project Javascript or Typescript: Javascript

### Describe the bug

We've faced a quite blocking crash when using dynamic components in our projet. Especially with rich texts fields causing a crash when trying to update the text after deleting a component above it.

### Steps to reproduce the behavior

1. Add 2 components in a dynamic zone, one must contain a richtext
2. Remove a component **before** the richtext one
3. Try to type in the richtext, you'll face an error _Woops! Something went wrong. Please, try again._
4. In the console there is an error `TypeError: Cannot read properties of undefined (reading 'icon')`

### Expected behavior

The richtext state should be updated with no error.

### Screenshots

image
[Screencast from 30-08-2023 10:46:06.webm](https://github.com/strapi/strapi/assets/60349428/834f5979-789e-4ce9-9090-7fa46223fd10)

### PR DESCRIPTION
**Pull request by jasperhartong**
### What does it do?

It fixes the stability of editing rich text in components within a Dynamic Zone;
- Ensures id's of newly added components don't conflict with existing id's
- Ensures that the editor wrapper is not rendered twice after adding new components above it

### Why is it needed?

When adding a new component in the Dynamic Zone it is assigned a Temporary Key. This temporary key is also used as id and key within the rendering loop of the components leading to conflicts. These conflicts show up as duplicate content.

E.g. Before this PR if there were 9 components in a dynamic zone, a newly added component would get id 10 assigned as temporary key. If there was already a same component with that id from the backend in the list it would show up as duplicate of that tinstance (until saving/ refreshing the editor). Now it will be assigned an id that is higher than any of the id's included in the array, this logic has been inspired by the same logic for stages: packages/core/admin/ee/admin/pages/SettingsPage/pages/ReviewWorkflows/reducer/index.js:209

The second fix is ensuring that the wrapper of CodeMIrror is not rendered twice when the parent component is re-rendered by cleaning it up if it ahs been instantiated before.

Eg. Whenever a component was added in a Dynamic Zone above a component containing the rich text editor, the wrapper would show up twice.

### How to test it?

- set up a dynamic zone with a component containing a rich text editor 
- Add one above and see the wrapper duplication and id generation fail

### Related issue(s)/PR(s)

- Fixes #17858 (breaks before PR, doesn't break after)

### CODE REVIEW

**Comment by jasperhartong** [comment]
Hi! Anything I can do to move this PR forward? Some of our editors actually loose data because of this issue.

**Comment by apppel** [comment]
What’s the status on this? Thank you.

File: packages/core/admin/admin/src/content-manager/components/Wysiwyg/Editor.js [hunk 1]
```diff
@@ -29,6 +29,11 @@ const Editor = forwardRef(
     const onChangeRef = useRef(onChange);
 
     useEffect(() => {
+      if (editorRef.current) {
+        // Ensure the editor and its wrapper are cleaned up whenever this view is re-rendered
+        // e.g. in case of re-ordering wysiwyg components in a DynamicZone
+        editorRef.current.toTextArea();
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/Wysiwyg/Editor.js:35]: Im not sure about this change, if placeholder or name are perceived to update then it'll try to focus itself, how does it work when there are many editors rendered at different places? Feels like a source of conflicts

File: packages/core/admin/admin/src/content-manager/utils/getMaxTempKey.js [hunk 2]
```diff
@@ -3,9 +3,13 @@ const getMaxTempKey = (arr) => {
...
   }
 
+  /**
+   * We check if there are ids or __temp_key__ because you may add an object to a list of objects
+   * already in the DB.
+   */
   const maxTempKey = Math.max.apply(
     Math,
-    arr.map((o) => o.__temp_key__ ?? 0)
+    arr.map((o) => o.id ?? o.__temp_key__ ?? 0)
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/utils/getMaxTempKey.js:12]: I don't understand the issue you're solving here or why we need to adjust the temp key in this way, could you re-explain please?

**Review by joshuaellis — CHANGES_REQUESTED** [review]
Some thoughts, sorry for the delay. There's a lot happening at the moment.

File: packages/core/admin/admin/src/content-manager/components/Wysiwyg/Editor.js (hunk 1 above)
Comment (jasperhartong) [reply packages/core/admin/admin/src/content-manager/components/Wysiwyg/Editor.js:35]: When many different editors are in place, there's no conflict as it relates to a specific reference. In my case I actually have about 10 rendered editors within different dynamic blocks.

About your remark on the placeholder and name changing leading to focusing issues:
- placeholder and name are indeed in the dependency array, but I'm under the assumption that these would not dynamically change after page load?
- the `toTextArea` method does not refocus anything. It just cleans up the CodeMirror instance, e.g. removing the wrapper element.

File: packages/core/admin/admin/src/content-manager/utils/getMaxTempKey.js (hunk 2 above)
Comment (jasperhartong) [reply packages/core/admin/admin/src/content-manager/utils/getMaxTempKey.js:12]: Say you have list of components in a dynamic zone with the following id's (the __temp_keys__ are set based on the index upon loading the data):

```
[
   {id: 2, __temp_key__: 0, __component: 'blog.simple' ....},
   {id: 3, __temp_key__: 1, __component: 'blog.simple'....}
]
```

Without the fix, when adding a new item to this zone, the `maxTempKey` function is called to retrieve a temporary client side id. This id, retrieved based on the highest `__temp_key__` in the list is eventually also set as id in the list:
```
[
   {id: 2, __temp_key__: 0, __component: 'blog.simple' ....},
   {id: 3, __temp_key__: 1, __component: 'blog.simple'....}
   {id: 2, __temp_key__: 2, __component: 'blog.simple'....}
]
```
This results in 2 items having the same id, which leads to weird, undefined behaviour.

The fix takes the `id` into account within `maxTempKey`, which in this case would result in:
```
[
   {id: 2, __temp_key__: 0, __component: 'blog.simple' ....},
   {id: 3, __temp_key__: 1, __component: 'blog.simple'....}
   {id: 4, __temp_key__: 4, __component: 'blog.simple'....}
]
```

Note that I didn't come up with this fix myself, it is copied from https://github.com/strapi/strapi/blob/d999ad0b00ff427885462051cdc96626b7946472/packages/core/admin/ee/admin/src/pages/SettingsPage/pages/ReviewWorkflows/reducer/index.js#L209-L214 as I assume the same key conflicting issue was present within the stages of review workflows.

---

Hop this clears things up :)

**Pull request closed without merging**